### DISCUSSION TITLE
How to prevent the closing of hd.dialog by clicking on arbitrary screen area

### DISCUSSION THREAD

**Discussion by cheak1974** [hyperdiv/hyperdiv#65] [Help]
Hi Everybody,
maybe I missed something in the docs but when I create a dialog I can close it either by clicking on the cross or on some arbitrary area outside the dialog. The second behavior is somehow uncool because you have absolutely no control of what is happening and you can't use the dialog to force a user decision. While playing around with dialog the following questions came up:

1. Is it possible to hide the Closing Cross and close the dialog with the help of buttons placed (and have the chance to exec code dependent on the users choice) in dialog or in the footer of the dialog?

2. How can I turn the dialog into a modal dialog (shoelace supports this) to prevent the dialog to be closed by clicking somewhere else than the wanted controls?

3. How can a dialog closing be done from buttons inside the dialog or even from outside (like a long running task). Reacting on the clicked event of the button and calling `dlgname.opened = False` isn't working.

regards
Chris

**Comment by mariusnita** [comment]
Hi! Yes, unfortunately there's no way currently to (a) hide the close button or (b) make the dialog unclosable when clicking the close button, clicking outside the dialog, or pressing the Escape key. I agree it would be useful in certain situations to prevent closing the dialog.

There is a way in Shoelace to make closing the dialog fully controllable by the programmer. See: https://github.com/shoelace-style/shoelace/discussions/471#discussioncomment-897940

I can investigate adding an interface to Hyperdiv's dialog to prevent closing the dialog until `opened` is explicitly set to `False` in Python.

I would have suggested doing this temporary workaround:
```py
d = hd.dialog(...)
...
# If dialog was closed, reopen it:
if d.visibility_changed and not d.opened:
    d.opened = True
```
But there's another Shoelace bug that prevents this from working: https://github.com/hyperdiv/hyperdiv/issues/54 😞 

But there's a jank workaround for that. You can re-open the dialog after a delay:
```py
# If dialog was closed, reopen it:
if d.visibility_changed and not d.opened:
    d.set_prop_delayed("opened", True, 1)
```
Setting a very short delay will also not work, due to the same bug mentioned above. The dialog has to close completely before it can be reopened.

You probably don't want to use this kind of pattern (while the dialog is briefly closed, the user could in theory interact with the UI underneath) but just wanted to show what I think is currently possible in terms of workarounds, until there's a robust solution.

As far as your point (3), setting `opened` to True/False should generally open/close the dialog. If there's a case where this doesn't work, you should file a bug with example code.

```py
def main():
    d = hd.dialog("My Dialog")
    with d:
        if hd.button("Close").clicked:
            d.opened = False
    if hd.button("Open").clicked:
        d.opened = True
```

**Reply by cheak1974** [reply]
Hi,

first of all thank you very much for your extensive explanation.

For me the reopening of the dialog while checking if the visibility has changed is not really a valid option. Depending on the contents (let's say maybe a video clip) this can result in effects that are not really nice.

The funny thing is that there is obviously an documented option in shoelace to avoid unwanted closing and the example works just fine.

[Shoelace prevent dialog from closing](https://shoelace.style/components/dialog#preventing-the-dialog-from-closing)


Maybe this will find it's way into hyperdiv (which I love a lot actually) one day. 

regards
Chris

**Reply by mariusnita** [reply]
Thanks, I plan to add that option to Hyperdiv.
